code/modules/research/anomaly/anomaly_refinery.dm

```diff
@@ -0,0 +1,300 @@
+	/// Here for the UI, tracks the amounts of reaction that has occured. 1 means valve opened but not reacted.
+	var/reaction_increment = 0
+
+/obj/machinery/research/anomaly_refinery/Initialize(mapload)
+	. = ..()
+	RegisterSignal(src, COMSIG_ATOM_INTERNAL_EXPLOSION, .proc/check_test)
+
+/obj/machinery/research/anomaly_refinery/assume_air(datum/gas_mixture/giver)
+	qdel(giver)
+	return null // Required to make the TTV not vent directly into the air.
+
+/**
+ * Determines how much explosive power (last value, so light impact theoretical radius) is required to make a certain anomaly type.
+ *
+ * Returns null if the max amount has already been reached.
+ *
+ * Arguments:
+ * * anomaly_type - anomaly type define
+ */
+/obj/machinery/research/anomaly_refinery/proc/get_required_radius(anomaly_type)
+	var/already_made = SSresearch.created_anomaly_types[anomaly_type]
+	var/hard_limit = SSresearch.anomaly_hard_limit_by_type[anomaly_type]
+	if(already_made >= hard_limit)
+		return //return null
+	// my crappy autoscale formula
+	// linear scaling.
+	var/radius_span = MAX_RADIUS_REQUIRED - MIN_RADIUS_REQUIRED
+	var/radius_increase_per_core = radius_span / hard_limit
+	var/radius = clamp(round(MIN_RADIUS_REQUIRED + radius_increase_per_core * already_made, 1), MIN_RADIUS_REQUIRED, MAX_RADIUS_REQUIRED)
+	return radius
+
+/obj/machinery/research/anomaly_refinery/attackby(obj/item/tool, mob/living/user, params)
+	if(active)
+		to_chat(user, span_warning("You can't insert [tool] into [src] while [p_theyre()] currently active."))
+		return
+	if(istype(tool, /obj/item/raw_anomaly_core))
+		if(inserted_core)
+			to_chat(user, span_warning("There is already a core in [src]."))
+			return
+		if(!user.transferItemToLoc(tool, src))
+			to_chat(user, span_warning("[tool] is stuck to your hand."))
+			return
+		var/obj/item/raw_anomaly_core/raw_core = tool
+		if(!get_required_radius(raw_core.anomaly_type))
+			say("Unfortunately, due to diminishing supplies of condensed anomalous matter, [raw_core] and any cores of its type are no longer of a sufficient quality level to be compressed into a working core.")
+		inserted_core = raw_core
+		to_chat(user, span_notice("You insert [raw_core] into [src]."))
+		return
+	if(istype(tool, /obj/item/transfer_valve))
+		if(inserted_bomb)
+			to_chat(user, span_warning("There is already a bomb in [src]."))
+			return
+		var/obj/item/transfer_valve/valve = tool
+		if(!valve.ready())
+			to_chat(user, span_warning("[valve] is incomplete."))
+			return
+		if(!user.transferItemToLoc(tool, src))
+			to_chat(user, span_warning("[tool] is stuck to your hand."))
+			return
+		inserted_bomb = tool
+		tank_to_target = inserted_bomb.tank_two
+		to_chat(user, span_notice("You insert [tool] into [src]"))
+		return
+	if(default_unfasten_wrench(user, tool))
+		return
+	if(default_deconstruction_screwdriver(user, "[base_icon_state]-off", "[base_icon_state]", tool))
+		return
+	if(default_deconstruction_crowbar(tool))
+		return
```

Comment:
Could you do these with the tool act procs

Reply:
want me to do the same for the other machines?

Reply:
Damn I like how it looks, so I went ahead and did it.